### PR TITLE
Contexts + CdnInfoContext & Gw2ClientContext

### PR DESCRIPTION
**Pull request by dlamkins**
Added base structure for new `ContextsService`.  Also added two contexts `CdnInfoContext` and `Gw2ClientContext` which demonstrate detecting the Client version of the application using the Mumble Link service to detect the current build ID and then the asset CDNs to check which client we are using based on that current build ID.

This provides some of the potential implementation currently being discussed in #62. 

Additionally, if accepted, this contains contexts that would allow us to more easily solve #50.

### CODE REVIEW

File: Blish HUD/GameServices/Contexts/CdnInfoContext.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+                if (cdnVars.Length == 5) {
+                    parsedSuccessfully &= int.TryParse(cdnVars[0], out _buildId);
+                    parsedSuccessfully &= int.TryParse(cdnVars[1], out _exeFileId);
+                    parsedSuccessfully &= int.TryParse(cdnVars[2], out _exeFileSize);
+                    parsedSuccessfully &= int.TryParse(cdnVars[3], out _manifestFileId);
+                    parsedSuccessfully &= int.TryParse(cdnVars[4], out _manifestFileSize);
+                }
+
+                if (!parsedSuccessfully) {
```
Comment (greaka) [inline Blish HUD/GameServices/Contexts/CdnInfoContext.cs]: minor bug: `parsedSuccessfully` can return the wrong result

```cs
bool parsedSuccessfully = false;

if (cdnVars.Length == 5) {
    parsedSuccessfully = true;
    ...
}
if (!parsedSuccessfully) {

File: Blish HUD/GameServices/Contexts/CdnInfoContext.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        private CdnInfo _standardCdnInfo;
+        private CdnInfo _chineseCdnInfo;
+
+        private int _loadCount = 0;
+
+        #region Context Loading
+
+        protected override void Load() {
+            GetCdnInfoFromCdnUrl(GW2_ASSETCDN_URL).ContinueWith((cdnInfo) => SetCdnInfo(ref _standardCdnInfo, cdnInfo.Result));
+            GetCdnInfoFromCdnUrl(GW2_CN_ASSETCDN_URL).ContinueWith((cdnInfo) => SetCdnInfo(ref _chineseCdnInfo, cdnInfo.Result));
```
Comment (greaka) [inline Blish HUD/GameServices/Contexts/CdnInfoContext.cs:62]: is `ContinueWith` safe to use when `GetCdnInfoFromCdnUrl` returns `null`?

File: Blish HUD/GameServices/Contexts/Gw2ClientContext.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+            if (contextAvailability == ContextAvailability.Available) {
+                int standardBuildId = standardCdnContextResult.Value.BuildId;
+
+                if (currentBuildId == standardBuildId) {
+                    return true;
+                }
+            }
+
+            return false;
```
Comment (greaka) [inline Blish HUD/GameServices/Contexts/Gw2ClientContext.cs]: This is probably preference, I would write that a little bit different for readability:

```cs
if (contextAvailability != ContextAvailability.Available)
    return false;

var standardBuildId = standardCdnContextResult.Value.BuildId;
return currentBuildId == standardBuildId;
```

File: Blish HUD/GameServices/Contexts/Gw2ClientContext.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+            if (contextAvailability == ContextAvailability.Available) {
+                int chineseBuildId = chineseCdnContextResult.Value.BuildId;
+
+                if (currentBuildId == chineseBuildId) {
+                    return true;
+                }
+            }
+
+            return false;
```
Comment (greaka) [inline Blish HUD/GameServices/Contexts/Gw2ClientContext.cs]: see comment above

**Review by greaka — CHANGES_REQUESTED** [review]
I like the propagation of the availability, even though c# makes it a bit weird to work with.
All in all a good concept. 


`Blish HUD/GameServices/Contexts/Gw2ClientContext.cs` needs a refresh on a game start, otherwise it will return the wrong client type whenever you switch the game client.
also, it will start logging that no version could be matched when the game build updates and we fetch a new mumble link build info

File: Blish HUD/GameServices/Contexts/CdnInfoContext.cs (hunk 1 above)
Comment (dlamkins) [reply Blish HUD/GameServices/Contexts/CdnInfoContext.cs]: Oh, good catch!  Is this too dirty of a fix?  https://dotnetfiddle.net/irafyW

File: Blish HUD/GameServices/Contexts/Gw2ClientContext.cs (hunk 3 above)
Comment (dlamkins) [reply Blish HUD/GameServices/Contexts/Gw2ClientContext.cs]: Yep!  Could even just be:

```cs
if (contextAvailability != ContextAvailability.Available)
    return false;

return currentBuildId == standardCdnContextResult.Value.BuildId;
```

That var isn't very helpful at that point, really.

File: Blish HUD/GameServices/Contexts/Gw2ClientContext.cs (hunk 4 above)
Comment (dlamkins) [reply Blish HUD/GameServices/Contexts/Gw2ClientContext.cs]: ✔ Agreed.

**Comment by dlamkins** [comment]
@greaka I responded to a couple of the items here so far.  Just FYI, this implementation was not completed.  It was more of a proof of concept that, if folks thought was decent enough, could later be finished to be the first implementation of it.  Refreshing some of the context info, handling failed HTTP requests, etc. would all be included if the concept itself looked good to you.  

I just didn't want to spend too much time on it unless folks confirmed they liked the implementation.  😄 

If you do like the implementation (ignoring the flaws of the current implemented contexts you've mentioned), or have any additional thoughts, could you let me know over in #62? 😃

File: Blish HUD/GameServices/Contexts/CdnInfoContext.cs (hunk 1 above)
Comment (greaka) [reply Blish HUD/GameServices/Contexts/CdnInfoContext.cs]: I would like to advise against doing that. the compiler should be able to optimize both solutions, but a normal developer would definitely oversee that you didnt write `==`

File: Blish HUD/GameServices/Contexts/Gw2ClientContext.cs (hunk 3 above)
Comment (greaka) [reply Blish HUD/GameServices/Contexts/Gw2ClientContext.cs]: I didnt want to change too much, but yes, I would omit the var as well

File: Blish HUD/GameServices/Contexts/CdnInfoContext.cs (hunk 1 above)
Comment (dlamkins) [reply Blish HUD/GameServices/Contexts/CdnInfoContext.cs]: Sounds good!  I'll put that change in tonight.

File: Blish HUD/GameServices/Contexts/Gw2ClientContext.cs (hunk 3 above)
Comment (dlamkins) [reply Blish HUD/GameServices/Contexts/Gw2ClientContext.cs]: Great - I've got this fixed in my local branch.  I'll push it tonight so we can start checking some of these items off.

File: Blish HUD/GameServices/Contexts/CdnInfoContext.cs (hunk 2 above)
Comment (dlamkins) [reply Blish HUD/GameServices/Contexts/CdnInfoContext.cs:62]: Yes, there shouldn't be any problems with this.  I tested last night with Fiddler to ensure various request failures would be handled correctly.

`SetCdnInfo` handles `null` without issue with its first check.  `ContinueWith` will pass the result regardless.

The local branch that I'll be pushing tonight has some additional logging to ensure the response status code and status reason are logged if the request fails for some reason.

**Comment by dlamkins** [comment]
- Resolved bug with CdnInfo where parsedSuccessfully could be true if the number of values was not 5. https://github.com/blish-hud/Blish-HUD/pull/106#discussion_r318001999
- Reduced complexity of "Is______ClientType" function.  https://github.com/blish-hud/Blish-HUD/pull/106#discussion_r318009487 & https://github.com/blish-hud/Blish-HUD/pull/106#discussion_r318009731
- Fixed bugs in CdnInfoContext as well as Gw2ClientContext.
- Contexts can now be unloaded.
- Gw2MumbleService now has an event "BuildIdChanged"
- OverlayService now, in addition to TacO, checks to see if the running client is the standard client or the Chinese client (moves the CornerIcons over if it is the Chinese client).

**Comment by dlamkins** [comment]
@greaka While I think the primary focus of this PR is the Contexts implementation, I think it'd be good to have Lei test the artifact that is created from this build and ensure the Blish HUD `CornerIcon` moves over as expected on a real Chinese version of the client.  I did a good amount of testing with Fiddler to fake the build ID that is returned and was able to confirm it should work.  Also ensured that it would fallback gracefully if something fails with the request (and logs are much more detailed with why it failed, now).

If you like this latest implementation, I can finish writing up the XML docs for it and mark it ready for a true review.

Once we get this merged, we can start adding things like the `FestivalContext`, which I am fairly excited about.  🎉

**Review by greaka — DISMISSED** [review]

**Comment by dlamkins** [comment]
@greaka Please re-review the implementation when you have a chance.  I believe I have implemented everything now and have put in XML documentation where appropriate.  I've also added a `State` member to `Context`s which indicates if the `Context` is loading, ready, etc.

To prevent anybody from unregistering a `Context`, I added a `ContextHandle` which can be used to expire a `Context`.  The `ContextHandle` associated with a `Context` type is returned when you call `RegisterContext`.  The full workflow would look something like:

```cs
// Register the context when the module is loaded
var myContextHandle = GameService.Contexts.RegisterContext(new ExampleContext());

// Expire context when the module is unloaded
myContextHandle.Expire();
```

Also, I received confirmation that this implementation has fixed Lei's problem!  🙂 
https://cdn.discordapp.com/attachments/568470477543178261/617894300327477249/jt_2019-09-02_09-30-59.png

**Review by greaka — APPROVED** [review]

**Comment by greaka** [comment]
I have multiple nitpicks in case the contexts are meant to be thread safe.
Having them threadsafe is probably a nice to have for now though

**Comment by dlamkins** [comment]
Agreed!

**Pull request merged**